<center>
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/images/SN_web_lightmode.png" width="300" alt="cognitiveclass.ai logo">
</center>

<h1 align=center><font size = 5>Assignment: Notebook for Graded Assessment</font></h1>


# Introduction

Using this Python notebook you will:

1.  Understand three Chicago datasets
2.  Load the three datasets into three tables in a SQLIte database
3.  Execute SQL queries to answer assignment questions


## Understand the datasets

To complete the assignment problems in this notebook you will be using three datasets that are available on the city of Chicago's Data Portal:

1.  <a href="https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Socioeconomic Indicators in Chicago</a>
2.  <a href="https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Public Schools</a>
3.  <a href="https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01">Chicago Crime Data</a>

### 1. Socioeconomic Indicators in Chicago

This dataset contains a selection of six socioeconomic indicators of public health significance and a “hardship index,” for each Chicago community area, for the years 2008 – 2012.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2](https://data.cityofchicago.org/Health-Human-Services/Census-Data-Selected-socioeconomic-indicators-in-C/kn9c-c2s2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 2. Chicago Public Schools

This dataset shows all school level performance data used to create CPS School Report Cards for the 2011-2012 school year. This dataset is provided by the city of Chicago's Data Portal.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t](https://data.cityofchicago.org/Education/Chicago-Public-Schools-Progress-Report-Cards-2011-/9xs2-f89t?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)

### 3. Chicago Crime Data

This dataset reflects reported incidents of crime (with the exception of murders where data exists for each victim) that occurred in the City of Chicago from 2001 to present, minus the most recent seven days.

A detailed description of this dataset and the original dataset can be obtained from the Chicago Data Portal at:

[https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2](https://data.cityofchicago.org/Public-Safety/Crimes-2001-to-present/ijzp-q8t2?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01&cm_mmc=Email_Newsletter-\_-Developer_Ed%2BTech-\_-WW_WW-\_-SkillsNetwork-Courses-IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork-20127838&cm_mmca1=000026UJ&cm_mmca2=10006555&cm_mmca3=M12345678&cvosrc=email.Newsletter.M12345678&cvo_campaign=000026UJ)


### Download the datasets

This assignment requires you to have these three tables populated with a subset of the whole datasets.

In many cases the dataset to be analyzed is available as a .CSV (comma separated values) file, perhaps on the internet. 

Use the links below to read the data files using the Pandas library. 

* Chicago Census Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Public Schools

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

* Chicago Crime Data

https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01

**NOTE:** Ensure you use the datasets available on the links above instead of directly from the Chicago Data Portal. The versions linked here are subsets of the original datasets and have some of the column names modified to be more database friendly which will make it easier to complete this assignment.


### Store the datasets in database tables

To analyze the data using SQL, it first needs to be loaded into SQLite DB.
We will create three tables in as under:

1.  **CENSUS_DATA**
2.  **CHICAGO_PUBLIC_SCHOOLS**
3.  **CHICAGO_CRIME_DATA**


Load the `pandas` and `sqlite3` libraries and establish a connection to `FinalDB.db`


In [2]:
import pandas as pd
import sqlite3
conn = sqlite3.connect('FinalDB.db')


Load the SQL magic module


In [1]:
%load_ext sql

Use `Pandas` to load the data available in the links above to dataframes. Use these dataframes to load data on to the database `FinalDB.db` as required tables.


In [43]:
#data_census = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCensusData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')
#data_crime = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoCrimeData.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')
data_public_school_1 = pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DB0201EN-SkillsNetwork/labs/FinalModule_Coursera_V5/data/ChicagoPublicSchools.csv?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDB0201ENSkillsNetwork20127838-2021-01-01')
#data_census.to_sql('Census Data',conn)
#data_crime.to_sql('Crime Data',conn)
data_public_school_1.to_sql('Public Schools Data1',conn)

Establish a connection between SQL magic module and the database `FinalDB.db`


In [4]:
%sql sqlite:///FinalDB.db

'Connected: @FinalDB.db'

You can now proceed to the the following questions. Please note that a graded assignment will follow this lab and there will be a question on each of the problems stated below. It can be from the answer you received or the code you write for this problem. Therefore, please keep a note of both your codes as well as the response you generate.


## Problems

Now write and execute SQL queries to solve assignment problems

### Problem 1

##### Find the total number of crimes recorded in the CRIME table.


In [6]:
df2 = pd.read_sql("SELECT * FROM 'Crime Data'",conn)
df2.describe()

index            ID      ARREST  ...        YEAR    LATITUDE   LONGITUDE
count  533.000000  5.330000e+02  533.000000  ...   533.00000  529.000000  529.000000
mean   266.000000  6.228301e+06    0.305816  ...  2008.34334   41.843429  -87.677818
std    154.008117  3.027454e+06    0.461185  ...     4.94940    0.082798    0.060073
min      0.000000  2.114900e+04    0.000000  ...  2001.00000   41.645796  -87.905227
25%    133.000000  3.420516e+06    0.000000  ...  2004.00000   41.772556  -87.718166
50%    266.000000  6.231712e+06    0.000000  ...  2008.00000   41.857737  -87.676733
75%    399.000000  8.875058e+06    1.000000  ...  2012.00000   41.902298  -87.634677
max    532.000000  1.127717e+07    1.000000  ...  2018.00000   42.021178  -87.528223

[8 rows x 13 columns]

### Problem 2

##### List community area names and numbers with per capita income less than 11000.


In [50]:
#df = pd.read_sql("SELECT COMMUNITY_AREA_NUMBER, COMMUNITY_AREA_NAME FROM 'Crime Data' WHERE PER_CAPITA_INCOME<11000",conn)
#print(df)

data_census


COMMUNITY_AREA_NUMBER  ... HARDSHIP_INDEX
0                     1.0  ...           39.0
1                     2.0  ...           46.0
2                     3.0  ...           20.0
3                     4.0  ...           17.0
4                     5.0  ...            6.0
..                    ...  ...            ...
73                   74.0  ...           16.0
74                   75.0  ...           30.0
75                   76.0  ...           24.0
76                   77.0  ...           19.0
77                    NaN  ...            NaN

[78 rows x 9 columns]

### Problem 3

##### List all case numbers for crimes involving minors?(children are not considered minors for the purposes of crime analysis) 


In [27]:
df = pd.read_sql("SELECT * FROM 'Crime Data' WHERE DESCRIPTION LIKE '%MINOR'",conn)
print(df)

   index       ID  ...  LONGITUDE                       LOCATION
0    514  3987219  ... -87.640899  (41.926268719, -87.640899336)
1    515  3266814  ... -87.584966   (41.72530099, -87.584965887)

[2 rows x 22 columns]


### Problem 4

##### List all kidnapping crimes involving a child?


In [30]:
df = pd.read_sql("SELECT * FROM 'Crime Data' WHERE PRIMARY_TYPE='KIDNAPPING' AND DESCRIPTION LIKE '%CHILD%'",conn)
print(df)

   index       ID  ...  LONGITUDE                       LOCATION
0    520  5276766  ... -87.750249  (41.874908413, -87.750249307)

[1 rows x 22 columns]


### Problem 5

##### List the kind of crimes that were recorded at schools. (No repetitions)


### Problem 6

##### List the type of schools along with the average safety score for each type.


In [49]:
df3 = pd.read_sql("SELECT * FROM 'Public Schools Data1' WHERE NAME_OF_SCHOOL LIKE'%Middle%'",conn)
print(df3)

   index  School_ID  ... Police_District                     Location
0     21     609780  ...              25  (41.91605519, -87.72177945)
1    152     610319  ...               9    (41.829365, -87.66778007)
2    162     610215  ...              10  (41.84098398, -87.70508585)
3    235     610532  ...               8  (41.79350627, -87.71139803)
4    338     610325  ...              14   (41.9339976, -87.70804359)
5    383     610018  ...               2  (41.80434474, -87.59039494)
6    411     610051  ...              25  (41.92053845, -87.75820003)
7    432     610395  ...               2  (41.83649005, -87.61837372)
8    509     610321  ...              17   (41.9520236, -87.72035833)
9    560     610336  ...               1  (41.84403809, -87.62859008)

[10 rows x 79 columns]


### Problem 7

##### List 5 community areas with highest % of households below poverty line


In [52]:
df4 = pd.read_sql("SELECT COMMUNITY_AREA_NAME FROM 'Census Data' ORDER BY PERCENT_HOUSEHOLDS_BELOW_POVERTY DESC LIMIT 5",conn)
print(df4)

  COMMUNITY_AREA_NAME
0           Riverdale
1         Fuller Park
2           Englewood
3      North Lawndale
4  East Garfield Park


### Problem 8

##### Which community area is most crime prone? Display the coumminty area number only.


In [54]:
df5 = pd.read_sql("SELECT * FROM 'Crime Data' ORDER BY DISTRICT ASC\
SELECT * FROM 'Crime Data' WHERE DISCTRICT = MAX",conn)
print(df5)

     index        ID  ...  LONGITUDE                       LOCATION
0       13   5343947  ... -87.639224  (41.868242722, -87.639223912)
1       21   6233210  ... -87.627758  (41.881597699, -87.627758082)
2       30   9859793  ... -87.624341   (41.879104092, -87.62434115)
3       40   1845995  ... -87.627715  (41.882851552, -87.627714909)
4       47   9575333  ... -87.634397  (41.880653112, -87.634396649)
..     ...       ...  ...        ...                            ...
528    425   2103277  ... -87.745055  (41.923183306, -87.745055091)
529    445  10268721  ... -87.742957  (41.931580848, -87.742956658)
530    474   5827212  ... -87.726850  (41.927235118, -87.726849666)
531    505   3588432  ... -87.736464   (41.92157476, -87.736464249)
532    509   8169210  ... -87.774454  (41.922537078, -87.774453593)

[533 rows x 22 columns]



Double-click **here** for a hint

<!--
Query for the 'community area number' that has most number of incidents
-->


### Problem 9

##### Use a sub-query to find the name of the community area with highest hardship index


### Problem 10

##### Use a sub-query to determine the Community Area Name with most number of crimes?


## Author(s)

<h4> Hima Vasudevan </h4>
<h4> Rav Ahuja </h4>
<h4> Ramesh Sannreddy </h4>

## Contribtuor(s)

<h4> Malika Singla </h4>
<h4>Abhishek Gagneja</h4>
<!--
## Change log

| Date       | Version | Changed by        | Change Description                             |
| ---------- | ------- | ----------------- | ---------------------------------------------- |
|2023-10-18  | 2.6     | Abhishek Gagneja  | Modified instruction set |
| 2022-03-04 | 2.5     | Lakshmi Holla     | Changed markdown.                   |
| 2021-05-19 | 2.4     | Lakshmi Holla     | Updated the question                           |
| 2021-04-30 | 2.3     | Malika Singla     | Updated the libraries                          |
| 2021-01-15 | 2.2     | Rav Ahuja         | Removed problem 11 and fixed changelog         |
| 2020-11-25 | 2.1     | Ramesh Sannareddy | Updated the problem statements, and datasets   |
| 2020-09-05 | 2.0     | Malika Singla     | Moved lab to course repo in GitLab             |
| 2018-07-18 | 1.0     | Rav Ahuja         | Several updates including loading instructions |
| 2018-05-04 | 0.1     | Hima Vasudevan    | Created initial version                        |
-->
## <h3 align="center"> © IBM Corporation 2023. All rights reserved. <h3/>
